File: LeetCode Challenge/LeetCode Youtube/go-leetcode/YuhLapTrinhVien/Vol(0-199)/20-Question/1502/1502.ts
/*
 * 🔢 Arithmetic Progression Checker
 * ==============================
 *
 * 📝 What is Arithmetic Progression?
 * - Each number differs from the next by a constant amount
 * - Example: [2, 4, 6, 8] → difference is always 2
 *
 * 🎯 Solution Steps:
 * 1. Find min and max numbers
 * 2. Calculate expected difference using (max-min)/(length-1)
 * 3. Check if all numbers in sequence exist
 */

function canMakeArithmeticProgression(arr: number[]): boolean {
  // Handle arrays with 0 or 1 elements
  if (arr.length <= 2) return true;

  // Find min and max
  let min = Math.min(...arr);
  let max = Math.max(...arr);

  // If all numbers are same
  if (min === max) return true;

  const n = arr.length;

  // Check if difference will be whole number
  if ((max - min) % (n - 1) !== 0) return false;

  // Calculate the expected difference
  const diff = (max - min) / (n - 1);

  // Use Set for O(1) lookup
  const numSet = new Set(arr);

  // Check if all numbers in sequence exist
  for (let i = 0; i < n; i++) {
    if (!numSet.has(min + i * diff)) {
      return false;
    }
  }

  return true;
}

// Test cases
console.log(canMakeArithmeticProgression([3, 5, 1])); // true: [1, 3, 5]
console.log(canMakeArithmeticProgression([1, 2, 4])); // false
console.log(canMakeArithmeticProgression([1, 1, 1, 1])); // true: all same

File: LeetCode Challenge/LeetCode Youtube/go-leetcode/YuhLapTrinhVien/Vol(0-199)/20-Question/1502/1502.go
package main

import "fmt"

/*
 * 📐 Arithmetic Progression Check Using Max-Min Formula
 * ================================================
 *
 * 🎯 Key Idea:
 * In an arithmetic progression, the difference between
 * any two consecutive numbers must be equal.
 *
 * 📝 Formula:
 * If a sequence is arithmetic, then:
 * (max - min) / (length - 1) = common difference
 *
 * 🌟 Examples:
 * 1. [2, 4, 6, 8]
 *    max = 8, min = 2, length = 4
 *    (8 - 2) / (4 - 1) = 2 (integer)
 *    Common difference is 2 ✅
 *
 * 2. [1, 2, 4]
 *    max = 4, min = 1, length = 3
 *    (4 - 1) / (3 - 1) = 1.5 (not integer)
 *    No common difference ❌
 */

func canMakeArithmeticProgression(arr []int) bool {
	// Handle small arrays
	if len(arr) <= 2 {
		return true
	}

	// Find min and max
	min, max := arr[0], arr[0]
	for _, num := range arr {
		if num < min {
			min = num
		}
		if num > max {
			max = num
		}
	}

	// If all numbers are same, it's an arithmetic progression
	if min == max {
		return true
	}

	// Calculate expected difference
	n := len(arr)
	if (max-min)%(n-1) != 0 {
		return false // If not divisible evenly, can't be arithmetic
	}
	diff := (max - min) / (n - 1)

	// Use map to check if all required numbers exist
	seen := make(map[int]bool)
	for _, num := range arr {
		seen[num] = true
	}

	// Check if all numbers in sequence exist
	for i := 0; i < n; i++ {
		if !seen[min+i*diff] {
			return false
		}
	}

	return true
}

func main() {
	// Test cases
	fmt.Println(canMakeArithmeticProgression([]int{3, 5, 1}))            // Output: true -> 4/2 = 2
	fmt.Println(canMakeArithmeticProgression([]int{1, 2, 4}))            // Output: false -> 3/2 = 1.5
	fmt.Println(canMakeArithmeticProgression([]int{2, 4, 6, 8, 10, 12})) // Output: true -> 10/5 = 2
}
